In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
sns.set(style='white', context='poster')

In [2]:
df=pd.read_excel('/home/suhananujum/Telecom customer Preprocessed Data.xlsx')
df.head()

Unnamed: 0  gender  SeniorCitizen  Partner  Dependents    tenure  \
0           0       0              0        1           0 -1.277445   
1           1       1              0        0           0  0.066327   
2           2       1              0        0           0 -1.236724   
3           3       1              0        0           0  0.514251   
4           4       0              0        0           0 -1.236724   

   PhoneService  MultipleLines  InternetService  OnlineSecurity  ...  \
0             0              1                0               0  ...   
1             1              0                0               2  ...   
2             1              0                0               2  ...   
3             0              1                0               2  ...   
4             1              0                1               0  ...   

   DeviceProtection  TechSupport  StreamingTV  StreamingMovies  Contract  \
0                 0            0            0                0         0   
1                 2            0            0                0         1   
2                 0            0            0                0         0   
3                 2            2            0                0         1   
4                 0            0            0                0         0   

   PaperlessBilling  PaymentMethod  MonthlyCharges  TotalCharges  Churn  
0                 1              2       -1.160323     -0.398608      0  
1                 0              3       -0.259629     -0.948762      0  
2                 1              3       -0.362660     -1.641883      1  
3                 0              0       -0.746535     -0.983710      0  
4                 1              2        0.197365     -1.235224      1  

[5 rows x 21 columns]

Our aim is to do a supervised UMAP on this dataset


In [3]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        7043 non-null   int64  
 1   gender            7043 non-null   int64  
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   int64  
 4   Dependents        7043 non-null   int64  
 5   tenure            7043 non-null   float64
 6   PhoneService      7043 non-null   int64  
 7   MultipleLines     7043 non-null   int64  
 8   InternetService   7043 non-null   int64  
 9   OnlineSecurity    7043 non-null   int64  
 10  OnlineBackup      7043 non-null   int64  
 11  DeviceProtection  7043 non-null   int64  
 12  TechSupport       7043 non-null   int64  
 13  StreamingTV       7043 non-null   int64  
 14  StreamingMovies   7043 non-null   int64  
 15  Contract          7043 non-null   int64  
 16  PaperlessBilling  7043 non-null   int64  


In [4]:
import umap

/home/suhananujum/.local/lib/python3.8/site-packages/umap/distances.py:1063: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.
  @numba.jit()
/home/suhananujum/.local/lib/python3.8/site-packages/umap/distances.py:1071: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.
  @numba.jit()
/home/suhananujum/.local/lib/python3.8/site-packages/umap/distances.py:1086: NumbaDeprecatio

In [5]:
data=df.iloc[:,:-1]
data.head()


Unnamed: 0  gender  SeniorCitizen  Partner  Dependents    tenure  \
0           0       0              0        1           0 -1.277445   
1           1       1              0        0           0  0.066327   
2           2       1              0        0           0 -1.236724   
3           3       1              0        0           0  0.514251   
4           4       0              0        0           0 -1.236724   

   PhoneService  MultipleLines  InternetService  OnlineSecurity  OnlineBackup  \
0             0              1                0               0             2   
1             1              0                0               2             0   
2             1              0                0               2             2   
3             0              1                0               2             0   
4             1              0                1               0             0   

   DeviceProtection  TechSupport  StreamingTV  StreamingMovies  Contract  \
0                 0            0            0                0         0   
1                 2            0            0                0         1   
2                 0            0            0                0         0   
3                 2            2            0                0         1   
4                 0            0            0                0         0   

   PaperlessBilling  PaymentMethod  MonthlyCharges  TotalCharges  
0                 1              2       -1.160323     -0.398608  
1                 0              3       -0.259629     -0.948762  
2                 1              3       -0.362660     -1.641883  
3                 0              0       -0.746535     -0.983710  
4                 1              2        0.197365     -1.235224

In [7]:
target=df.iloc[:,-1]
target.head()


0    0
1    0
2    1
3    0
4    1
Name: Churn, dtype: int64

In [8]:
target.value_counts(normalize=False, sort=True, ascending=False, bins=None, dropna=True)

0    5174
1    1869
Name: Churn, dtype: int64

There are 5174 people labelled 0 so for the time being we will assume that this was a low churn i.e these people renewed their subscriptions

In [9]:
classes = [
    '0',
    '1']

In [ ]:
from umap import UMAP
from sklearn.pipeline import Pipeline
from sklearn.svm import SVC
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import RandomizedSearchCV
from sklearn.preprocessing import StandardScaler
import numpy as np
from sklearn.datasets import load_iris
import matplotlib.pyplot as plt
%matplotlib inline

# get toy data
iris = load_iris()
X, y = df.iloc[:,:-1], df.iloc[:,-1]

# cross validation strategy
cv = StratifiedKFold(n_splits=5).get_n_splits(X, y)

# UMAP followed by classification with a linear SVM
grid_pipeline = Pipeline([('scale', StandardScaler()),
                          ('umap', UMAP(random_state=42)), 
                          ('svc', SVC(random_state=42))])

params_grid_pipeline = {
    'umap__n_neighbors': np.arange(2,17),
    'umap__n_components': [2], # for 2D projection
    'umap__min_dist': np.linspace(0, 1, 10),
    'umap__metric': ['euclidean', 
                     'manhattan', 
#                      'chebyshev', 
#                      'minkowski',
#                      'canberra',
#                      'braycurtis',
#                      'haversine',
#                      'mahalanobis',
#                      'wminkowski',
#                      'seuclidean'
                    ],
    'svc__C': np.logspace(-4, 4, num=5),
    'svc__kernel': ['linear', 'rbf']
}

# https://stats.stackexchange.com/questions/160479/practical-hyperparameter-optimization-random-vs-grid-search
clf_pipeline = RandomizedSearchCV(grid_pipeline,
                                  params_grid_pipeline,
                                  scoring='accuracy',
                                  cv=cv,
                                  verbose=2,
                                  random_state=42, 
                                  n_iter=60, 
                                  n_jobs=-1)

clf_pipeline.fit(X, y)

print(clf_pipeline.best_params_)

# drop SVC and keep parameters for UMAP only
best_params = {key.replace('umap__',''):value \
               for key, value in clf_pipeline.best_params_.items() if 'umap__' in key}

umap_pipeline = Pipeline([('scale', StandardScaler()), 
                          ('umap', UMAP(random_state=42, **best_params))])

standard_embedding = umap_pipeline.fit_transform(X)

# plot results
plt.scatter(standard_embedding[:, 0], 
            standard_embedding[:, 1], 
            c=y,
            s=10,
            cmap='jet')

Fitting 5 folds for each of 60 candidates, totalling 300 fits
